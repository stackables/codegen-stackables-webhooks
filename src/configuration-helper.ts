interface ConfigurationOptions {
    generatedFile: string
    accountSlug: string
    introspectionToken: string
}

export function getConfiguration(opts: ConfigurationOptions) {
    const schemaUrl = `https://data.stackables.io/${opts.accountSlug}?introspection=${opts.introspectionToken}`
    return {
        schema: schemaUrl,
        extensions: {
            codegen: {
                generates: {
                    [opts.generatedFile]: {
                        documents: [
                            {
                                [schemaUrl]: {
                                    loader: 'codegen-stackables-webhooks',
                                },
                            },
                        ],
                        plugins: ['typescript', 'typescript-operations', 'stackables-webhooks'],
                        config: {
                            preResolveTypes: true,
                            onlyOperationTypes: true,
                            fragmentRegistryName: 'StackablesEvents',
                            fragmentRegistryDirective: 'register'
                        }
                    }
                }
            }
        }
    }
}
