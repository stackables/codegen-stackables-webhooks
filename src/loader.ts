import { parse } from 'graphql';
import { GraphQLClient, gql } from 'graphql-request'

export default async function (endpoint: string) {

  const graphQLClient = new GraphQLClient(endpoint)

  const query = gql`
    query CodeGenWebhooks {
      Webhook {
        list {
          name
          query
          entity: entity2 {
            id
          }
        }
      }
    }
  `

  const data = await graphQLClient.request(query)
  const fragments = data.Webhook.list;

  return parse(fragments.map((f: any) => `fragment ${f.name} on ${f.entity.id} @register ${f.query}`).join('\n'));
};