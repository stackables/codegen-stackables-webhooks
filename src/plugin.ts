import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { capitalCase } from 'change-case-all';
import { concatAST, DirectiveNode, DocumentNode, GraphQLSchema, visit } from 'graphql';

interface NamedOperationsObjectPluginConfig {
  fragmentRegistryName?: string
  fragmentRegistryDirective?: string
  omitOperationSuffix?: boolean
}

function filterByDirective(directives: readonly DirectiveNode[] | undefined, needed: string | undefined) {
  // if no restriction specified then list all
  if (!needed) return true

  // TODO: Investigate under what circumstances can it actually be undefined?
  /* istanbul ignore next */
  if (!directives) return false

  for (let index = 0; index < directives.length; index++) {
    const directive = directives[index]

    if (directive.name.value === needed) {
      return true
    }
  }

  // if nothing matched then we ignore this fragment
  return false
}

export const plugin: PluginFunction<NamedOperationsObjectPluginConfig, string> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: NamedOperationsObjectPluginConfig
) => {

  const fragmentRegistryName = config.fragmentRegistryName || 'FragmentRegistry'

  const docs = documents.map(v => v.document) as DocumentNode[]
  const allAst = concatAST(docs);
  const fragments: string[] = []

  visit(allAst, {
    FragmentDefinition: node => {
      if (filterByDirective(node.directives, config.fragmentRegistryDirective)) {
        fragments.push(node.name.value)
      }
    },
  });

  const suffix = (config.omitOperationSuffix) ? '' : 'Fragment'

  const source = []

  // Wrapper type
  source.push(`export type StackablesEnvelope<T> = { 
    version: "0.1"
    payload: T
    params: any
    callback: {
      url: string
      token: string
    }
  }`)

  // Handler functions
  fragments.map(fragment => {
    source.push(`export type ${fragment}Handler = StackablesEnvelope<${fragment}${suffix}>`)
  })

  source.push(`export type ${fragmentRegistryName} = {
${fragments.map(fragment => `  'stackables.webhook.${fragment}': ${fragment}Handler`).join('\n')}
  }`)

  return source.join('\n')
};
