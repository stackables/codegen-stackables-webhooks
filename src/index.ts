export { getConfiguration } from './configuration-helper'
export { plugin } from './plugin'
import loader from './loader'

export default loader