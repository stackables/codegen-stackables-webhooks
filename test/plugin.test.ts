import { Types } from '@graphql-codegen/plugin-helpers'
import { plugin } from '../src'
import { loadDocuments } from '@graphql-tools/load'

test('Plugin generated wrapped types for provided documents with correct decorator', async () => {

    // Schema is not used for index generation
    const schema = undefined as any

    // Generate test documents
    const fragments = await loadDocuments(`
    
    fragment IgnoreThisOne on TestType {
        id
    }

    fragment Webhook1 on TestType @generate1 {
        id
        name
    }

    fragment Webhook2 on TestType @generate2 {
        id
    }

    `, {} as any)

    const output1 = plugin(schema, fragments, {
        fragmentRegistryDirective: 'generate1',
        omitOperationSuffix: false
    })
    expect(output1).toMatchSnapshot()

    const output2 = plugin(schema, fragments, {
        fragmentRegistryDirective: 'generate2',
        omitOperationSuffix: true
    })
    expect(output2).toMatchSnapshot()

    const output3 = plugin(schema, fragments, {})
    expect(output3).toMatchSnapshot()
})