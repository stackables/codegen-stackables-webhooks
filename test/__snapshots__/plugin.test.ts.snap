// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin generated wrapped types for provided documents with correct decorator 1`] = `
"export type StackablesEnvelope<T> = { 
    version: \\"0.1\\"
    payload: T
    params: any
    callback: {
      url: string
      token: string
    }
  }
export type Webhook1Handler = StackablesEnvelope<Webhook1Fragment>
export type FragmentRegistry = {
  'stackables.webhook.Webhook1': Webhook1Handler
  }"
`;

exports[`Plugin generated wrapped types for provided documents with correct decorator 2`] = `
"export type StackablesEnvelope<T> = { 
    version: \\"0.1\\"
    payload: T
    params: any
    callback: {
      url: string
      token: string
    }
  }
export type Webhook2Handler = StackablesEnvelope<Webhook2>
export type WebhookMap = {
  'stackables.webhook.Webhook2': Webhook2Handler
  }"
`;

exports[`Plugin generated wrapped types for provided documents with correct decorator 3`] = `
"export type StackablesEnvelope<T> = { 
    version: \\"0.1\\"
    payload: T
    params: any
    callback: {
      url: string
      token: string
    }
  }
export type IgnoreThisOneHandler = StackablesEnvelope<IgnoreThisOneFragment>
export type Webhook1Handler = StackablesEnvelope<Webhook1Fragment>
export type Webhook2Handler = StackablesEnvelope<Webhook2Fragment>
export type FragmentRegistry = {
  'stackables.webhook.IgnoreThisOne': IgnoreThisOneHandler
  'stackables.webhook.Webhook1': Webhook1Handler
  'stackables.webhook.Webhook2': Webhook2Handler
  }"
`;
